### PR TITLE
ENH: optimize StringEncoder

### PR DESCRIPTION
**Pull request by GaelVaroquaux**
For memory and speed:

- Significant memory improvement (I did not measure)
- 1.5 speedup on one task

### CODE REVIEW

**Comment by GaelVaroquaux** [comment]
FYI, the examples can be seen here: https://output.circle-artifacts.com/output/job/f25f287d-9bba-4153-977a-bbd5fbc2f3ca/artifacts/0/doc/auto_examples/02_text_with_string_encoders.html#sphx-glr-auto-examples-02-text-with-string-encoders-py
And I'm quite happy that with this PR StringEncoder is now faster than MinHashEncoder on this example :v:

File: skrub/_string_encoder.py [hunk 1]
```diff
@@ -133,12 +133,17 @@ def fit_transform(self, X, y=None):
...
 
-        if (min_shape := min(X_out.shape)) >= self.n_components:
-            self.tsvd_ = TruncatedSVD(n_components=self.n_components)
+        if (min_shape := min(X_out.shape)) > self.n_components:
+            self.tsvd_ = TruncatedSVD(
+                n_components=self.n_components, algorithm="arpack"
+            )
             result = self.tsvd_.fit_transform(X_out)
+        elif X_out.shape[1] == self.n_components:
+            result = X_out.toarray()
```
Comment (jeromedockes) [inline skrub/_string_encoder.py:146]: does the 'arpack' algorithm not like it when p == n_components? or is skipping the tsvd in that case an optimization?  if the latter, I guess the case where the number of discovered ngrams is exactly equal to `n_components` might be too rare to warrant it? and we might want a test for that branch

File: skrub/_string_encoder.py [hunk 2]
```diff
@@ -152,6 +157,8 @@ def fit_transform(self, X, y=None):
             # Therefore, self.n_components_ below stores the resulting
             # number of dimensions of result.
             result = X_out[:, : self.n_components].toarray()
+            result = result.copy()
```
Comment (jeromedockes) [inline skrub/_string_encoder.py:160]: I suppose the `copy()` is that otherwise the reference to the slice would prevent `X_out` from being garbage collected; might be worth a short comment

**Review by jeromedockes — COMMENTED** [review]
LGTM but it would be nice to cover the shape[1] == n_components case in a test, LMK if you don't have time now I can push a small commit for it

**Comment by GaelVaroquaux** [comment]
> LGTM but it would be nice to cover the shape[1] == n_components case in a test, LMK if you don't have time now I can push a small commit for it

It would be helpful. Thanks!

Let's also wait for a systematic evaluation before merging

File: skrub/_string_encoder.py (hunk 1 above)
Comment (GaelVaroquaux) [reply skrub/_string_encoder.py:146]: It's both, actually.

I think that needed this for the tests to pass. At least my first implementation hit a corner case in the tests (good tests!)

**Review by jeromedockes — APPROVED** [review]
LGTM, thanks!

**Review by Vincent-Maladiere — APPROVED** [review]
Looks good!

**Comment by GaelVaroquaux** [comment]
@rcap107 : super useful. Thanks!!!
